### ISSUE TITLE
Duplicate guarded transition allowed in Umple

### ISSUE THREAD

**Issue by opeyemiAdesina** [umple/umple#760]
class CarTransmission {

  Boolean driveSelected;
  state {
    neutral {
      selectReverse -> reverse;
      selectDrive -> drive;
      selectFirst -> first;
      selectSecond -> second;
    }

```
reverse {
  selectNeutral -> neutral;
}

drive {
  selectNeutral -> neutral;
  selectFirst -> first;
  selectSecond -> second;

  first {
    reachSecondSpeed [driveSelected]  -> second;      
    reachSecondSpeed [driveSelected]  -> second;  //duplicate guarded transition 
  }
  second {  }
}
```

  }
}

**Comment by vahdat-ab** [comment]
I think we need to build a truth table for conditions and make sure they are not the same. This concept can be more complicated when we have fix values participated in the boolean expression or one of the conditions are reduced. 
I believe this algorithm is really important for Umple in order to guarantee non-deterministic state machines. I have done a small search (as part of my own work) but couldn't find a concrete algorithm for this purpose.

**Comment by TimLethbridge** [comment]
Right. So duplicate guarded transitions are OK as long as there are no situations where two of the guards can be true at the same time. If two guards are identical, as in the example at the start of this issue then there should be a warning. Then there should be a different warning if they are not the same but are not mutually exclusive. There should be no warning if they are mutually exclusive.

**Comment by opeyemiAdesina** [comment]
A warning is sufficient for the case presented in the issue. However, I have implemented a temporary solution to prevent such duplicates in the list returned by "*.getAllTransitions()" as it guarantees correct output for nuXmv pending the time the issue is resolved.

**Comment by vahdat-ab** [comment]
Consider the following simple example:

```
class A{
    Boolean x;
    Boolean y;
    status{
        on{
            turnOff[x<y] ->off; //n1
            turnOff[y>x] ->off; //n2
            turnOff[x<y] -> on; //n3
        }
        off{
            turnOn[x>y] -> on; //f1
            turnOn[y<=x] -> on; //f2
            turnOn[y==x] -> off;//f3
        }
    }
}
```

There is no conflict between `n1 and n2` because they go to the same state. Therefore, there is no worry. We may consider to just warn the designer about his/her design.

There is conflict obviously between `(n1 and n3)`  and `(n2 and n3)`. I think we should generate an error because it makes Umple state machines non-deterministic, except if would like to support non-deterministic state machines. In that case, we must provide a mechanism to deal those cases, such as prioritization.

There is another non-deterministic case between `f2 and f3` .

**Comment by tlaport4** [comment]
Is there a final verdict on whether we should be generating an error or just a warning in the cases outlined in Vahdat's last post?

**Comment by TimLethbridge** [comment]
If there is no guard, it is a warning if there are two identical events. We get:
45 Warning on line 10 : Transition xxxxx will be ignored due to a unguarded duplicate event.
If there *are* guards and both event and guards are identical, then logically it should also be a warning. The duplicate is ignored. It should work like error 45 does now. Message would be 
Transition xxxxx will be ignored to to duplicate events with identical guards.

**Comment by tlaport4** [comment]
In PR https://github.com/umple/umple/pull/938, I added logic to ignore transitions that have the same event and equivalent guard as another transition. The warning message displayed is "Guarded Transition xxxxx will be ignored due to a duplicate event with an equivalent guard".

Also works for auto transitions that have equivalent guards.

The logic for determining whether two guards are equivalent will do most basic comparison between the guards. For example, two guards like [(x + y) < z] and [z >= (y+x)] would be confirmed as equivalent. Also, works for more complicated guards that use negation and constraintBody's (parenthesis -- OPEN_ROUND_BRACKET [[constraint]] CLOSE_ROUND_BRACKET). See PR

**Comment by vahdat-ab** [comment]
@tlaport4  thank.
Consider the following example
```
class A{
  sm{
    on{
      [x]-> off;
      -> on;
    }
    off{
      
    }
    
  }
}
```
There are two warning messages for one transition.
```
Warning 54 on line 7 of file "test.ump":
Transition  will be ignored due to an unguarded duplicate event
Warning 55 on line 7 of file "test.ump":
Transition  in super-state will be ignored due to unguarded duplicate events in all sub-states
```
Check all locations and test cases related to warning 54 and see if your algorithm can satisfy them. Then, you can keep yours and delete others. It's better to keep one which is more comprehensive.

**Comment by vahdat-ab** [comment]
Another example
```
class A{
  sm{
    on{
      [x>y]-> off;
      [y<x]-> on;
    }
    off{  } 
  }
```
This should not be allowed.

**Comment by tlaport4** [comment]
Hmmm yeah that's right.

When I wrote I for some reason I thought the equivalent of > would be <=, but that's the opposite. I'll quickly update it.

**Comment by tlaport4** [comment]
I updated the CR and all the tests just passed.

**Comment by vahdat-ab** [comment]
You algorithm cannot detect the issue related to f2 and f3 in the above [example ](https://github.com/umple/umple/issues/760#issuecomment-189479114)

**Comment by tlaport4** [comment]
Those aren't necessarily equivalent though. They're only equivalent when x = y. What should it do? Should either one be ignored?

**Comment by vahdat-ab** [comment]
This is a dynamic behavior and from my perspective, it must be notified to the modeler. However, maybe @TimLethbridge  and @opeyemiAdesina want to comment more about this case.
This clearly shows that the state machine has this potential to be in two states. However, it might never happen because of a restriction on the values of x and y (e.g. OCL constraint). For example, someone might define a constraint in which it is said x and y must not be equal.

**Comment by TimLethbridge** [comment]
This is a case where one guard is a sub-case of another. We can't easily compute all these cases logically (i.e. analyse all truth tables). In this case one could compute easily that one is a sub-case, but cases get complex very quickly.  I say we leave this unless we are going to go 'whole-hog' and try to compute all the cases that are computable statically (this is a static case).

I don't call this non-determinism by the way. In Umple, order of declaration determines which case applies. It is only nondeterministic if it can change. But it is under-specified what happens in the presence of traits, and with mixins the order can be hard to be sure about.

**Comment by TimLethbridge** [comment]
Maybe this issue can be closed, as it is partly solved. But more testing would be useful
